File: 02.每日一题/单词距离/solution.cpp
//
// Created by lifan on 2022/5/27.
//

#include <map>
#include <string>
#include <vector>
#include <cmath>
#include <iostream>

using namespace std;

class Solution {

 public:
  int findClosest1(vector<string>& words, string word1, string word2) {
      int idx1 = -1, idx2 = -1;
      int min_dist = INT_MAX;

      for (int i = 0; i < words.size(); ++i) {
          if (words[i] == word1) {
              idx1 = i;
          }
          if (words[i] == word2) {
              idx2 = i;
          }

          if (idx1 != -1 && idx2 != -1) {
              min_dist = min(min_dist, abs(idx1 - idx2));
          }
      }

      return min_dist;
  }

  int findClosest(vector<string>& words, string word1, string word2) {
      map<string, vector<int>> word_index_map;

      for (int idx = 0; idx < words.size(); idx ++) {
          if ((word_index_map.count(words[idx])) == 0) {
              word_index_map[words[idx]] = vector<int>();
          }

          word_index_map[words[idx]].push_back(idx);
      }

      vector<int> index1 = word_index_map[word1];
      vector<int> index2 = word_index_map[word2];

      if (index1.empty() || index2.empty()) {
          return -1;
      }

      int i = 0, j = 0;
      int min_dist = INT_MAX;

      while (i < index1.size() && j < index2.size()) {
          min_dist = min(min_dist, abs(index1[i] - index2[j]));
          if (index1[i] < index2[j]) {
              i++;
          } else {
              j++;
          }
      }

      return min_dist;
  }
};

int main(int argc, char *argv[])
{
    Solution s;
    vector<string> words = {"1", "2", "3", "4"};
    int ans = s.findClosest(words,"5", "6");
    cout << ans;

    return 0;
}

File: 02.每日一题/单词距离/solution.py
from typing import List


class Solution:
    # 遍历一次
    def findClosest(self, words: List[str], word1: str, word2: str) -> int:
        idx1 = -1
        idx2 = -1
        min_dist = -1

        for idx, word in enumerate(words):
            if word == word1:
                idx1 = idx
            if word == word2:
                idx2 = idx

            if idx1 != -1 and idx2 != -1:
                if min_dist == -1:
                    min_dist = abs(idx1 - idx2)
                    continue
                min_dist = min(min_dist, abs(idx1 - idx2))

        return min_dist

    # 带缓存
    def findClosest(self, words: List[str], word1: str, word2: str) -> int:
        word_index_map = {}
        for idx, word in enumerate(words):
            if word in word_index_map:
                word_index_map[word].append(idx)
            else:
                word_index_map[word] = [idx]

        index1 = word_index_map[word1]
        index2 = word_index_map[word2]

        if len(index1) == 0 or len(index2) == 0:
            return -1

        idx1 = 0
        idx2 = 0
        min_dist = -1

        # 双指针移动的方法
        while idx1 < len(index1) and idx2 < len(index2):
            if min_dist == -1:
                min_dist = abs(index1[idx1] - index2[idx2])
            else:
                min_dist = min(min_dist, abs(index1[idx1] - index2[idx2]))

            if index1[idx1] < index2[idx2]:
                idx1 += 1
            else:
                idx2 += 1

        return min_dist
